client/src/components/mapping/FhirMappingPanel/FhirResourceTree/index.tsx

```diff
@@ -84,137 +67,122 @@ const FhirResourceTree = ({ onClickCallback }: Props) => {
-      secondaryLabel: secondaryLabel(path, childNodes),
-      nodeData: path
+      id: attribute.tail,
+      secondaryLabel: secondaryLabel(attribute, childNodes),
+      nodeData: attribute,
+      label: ''
     };
+    node.label = (
+      <NodeLabel
+        attribute={attribute}
+        addNodeCallback={() => addNodeToArray(node)}
+        deleteNodeCallback={() => deleteNodeFromArray(node, parentArray!)}
+      />
+    );
+    return node;
   };
 
   const deleteNodeFromArray = (
-    stateNodes: ITreeNode<Node>[],
-    path: Node
-  ): ITreeNode<Node>[] => {
+    deletedNode: ITreeNode<Attribute>,
+    arrayNode: ITreeNode<Attribute>
+  ) => {
     // First, we delete all the corresponding attributes in DB
+    const deleted = deletedNode.nodeData!;
+    const array = arrayNode.nodeData!;
     deleteAttributes({
       variables: {
-        startsWith: path.serialize()
+        startsWith: deleted.path
       }
     });
     // Same in Redux store
-    dispatch(removeAttributesFromMap(path.serialize()));
-
-    // Then we find the parent from which we want to remove
-    // the child node.
-    const parent = findNode(stateNodes, path.parent!);
-    if (!parent || !parent.childNodes) return stateNodes;
+    dispatch(removeAttributesFromMap(deleted.path));
+    // and remove the item in the parent attribute as well
+    array.removeItem(deleted);
 
-    if (parent.childNodes.length === 1) {
+    if (arrayNode.childNodes?.length === 1) {
       // If only one child is left, we simply rebuild an empty one
-      const childNode = new Node(path, path.definition, 0);
-      parent.childNodes = [
-        createNode(childNode, [], undefined, () =>
-          setNodes(nodes => deleteNodeFromArray(nodes, childNode))
-        )
-      ];
+      arrayNode.childNodes = [newItemNode(arrayNode)];
     } else {
       // else we remove the childNode from the parent
-      parent.childNodes = parent.childNodes.filter(
-        child => child.id !== path.tail()
+      arrayNode.childNodes = arrayNode.childNodes?.filter(
+        child => child.id !== deleted.tail
       );
     }
 
-    return stateNodes;
+    setNodes(nodes => [...nodes]);
   };
 
-  const addNodeToArray = (
-    stateNodes: ITreeNode<Node>[],
-    path: Node,
-    contentChildren: ITreeNode<Node>[]
-  ): ITreeNode<Node>[] => {
-    const parent = findNode(stateNodes, path);
-    if (!parent || !parent.childNodes) return stateNodes;
-
-    // find the last index used in the children and increment it.
-    const nextIndex =
-      parent.childNodes[parent.childNodes.length - 1].nodeData?.index! + 1;
-
-    // build a new child node
-    const childNode = new Node(path, path.definition, nextIndex);
-    const newNode = createNode(
-      childNode,
-      genTreeLevel(contentChildren || [], childNode),
-      undefined,
-      () => setNodes(nodes => deleteNodeFromArray(nodes, childNode))
+  const newItemNode = (arrayNode: ITreeNode<Attribute>, index?: number) => {
```

Comment:
Do you think the Attribute class could have a method addChild, createChild or something like that?
It could be called on a Attribute which is an array and return a new child or add it directly to it children.

Reply:
the Attribute class already has such a method, but the newItemNode function is about adding an ITreeNode<Attribute> so that wouldn't fit into the Attribute class. Did I get your point ?

Reply:
I'm not sure. I just wanted to avoid the `array.definition.max = '1'/'*';` hack to be done here

Reply:
I think that's better indeed, i'll try that!